### PR TITLE
derive channel pool size from cpu count

### LINKED ISSUE #871: Default channel pool size should be based on CPU count
The default channel pool size should use `std::thread::hardware_concurrency()` to auto configure the channel pool.

### PR DESCRIPTION
**Pull request by manish-qlogic**
This fixes #871.

---
This change is [Reviewable](https://reviewable.io/reviews/googlecloudplatform/google-cloud-cpp/956)

### CODE REVIEW

File: google/cloud/bigtable/client_options.cc [hunk 1]
```diff
@@ -32,9 +33,15 @@ namespace google {
 namespace cloud {
 namespace bigtable {
 inline namespace BIGTABLE_CLIENT_NS {
+inline std::size_t CalculateDefaultConnectionPoolSize() {
```
[the remark this answers is not in the document]
Comment (manish-qlogic) [reply google/cloud/bigtable/client_options.cc]: Java currently sets it to 2 times of the cpu for DataClient.
https://github.com/GoogleCloudPlatform/google-cloud-java/blob/c1fb3904c648d4db17b7aa8bbcd970d0e0374bb6/google-cloud-clients/google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/EnhancedBigtableStubSettings.java#L228

https://github.com/googleapis/gax-java/blob/4a3ab1e46ca7f0a8589a889370d7b1eb1f3189c1/gax-grpc/src/main/java/com/google/api/gax/grpc/InstantiatingGrpcChannelProvider.java#L395

and, channels per cpu is only set for data client where as no limit is set for Admin and InstanceAdmin (Defaults to single channel)

https://github.com/GoogleCloudPlatform/google-cloud-java/blob/c1fb3904c648d4db17b7aa8bbcd970d0e0374bb6/google-cloud-clients/google-cloud-bigtable-admin/src/main/java/com/google/cloud/bigtable/admin/v2/stub/BigtableInstanceAdminStubSettings.java#L329

So for default pool size, either we can set minimum size (as per GO) or rely on cpu count (as per JAVA) with a caveat (std::thread::hardware_concurrency() can return 0). 
I will add comment and add a constants BIGTABLE_CLIENT_DEFAULT_CHANNELS_PER_CPU (4) to control the multiplicity.

[the remark this answers is not in the document]
Comment (manish-qlogic) [reply google/cloud/bigtable/client_options.cc]: fixed

[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/bigtable/client_options.cc]: Thanks for the changes, can you set `BIGTABZLE_CLIENT_DEFAULT_CHANNELS_PER_CPU` to 2 like Java?

[the remark this answers is not in the document]
Comment (manish-qlogic) [reply google/cloud/bigtable/client_options.cc]: fixed.

**Comment by manish-qlogic** [comment]
PTAL

**Pull request merged**